crypto/bp/bp_fp2.c

```diff
@@ -0,0 +1,605 @@
+}
+
+int FP2_add(const BP_GROUP *group, FP2 *r, const FP2 *a, const FP2 *b)
+{
+    if (!BN_mod_add_quick(r->f[0], a->f[0], b->f[0], group->field))
+        return 0;
+    if (!BN_mod_add_quick(r->f[1], a->f[1], b->f[1], group->field))
+        return 0;
+    return 1;
+}
+
+int FP2_dbl(const BP_GROUP *group, FP2 *r, const FP2 *a)
+{
+    if (!BN_mod_lshift1_quick(r->f[0], a->f[0], group->field))
+        return 0;
+    if (!BN_mod_lshift1_quick(r->f[1], a->f[1], group->field))
+        return 0;
+    return 1;
+}
+
+int FP2_sub(const BP_GROUP *group, FP2 *r, const FP2 *a, const FP2 *b)
+{
+    if (!BN_mod_sub_quick(r->f[0], a->f[0], b->f[0], group->field))
+        return 0;
+    if (!BN_mod_sub_quick(r->f[1], a->f[1], b->f[1], group->field))
+        return 0;
+    return 1;
+}
+
+int FP2_neg(const BP_GROUP *group, FP2 *r, const FP2 *a)
+{
+    if (!BN_sub(r->f[0], group->field, a->f[0]))
+        return 0;
+    if (!BN_sub(r->f[1], group->field, a->f[1]))
+        return 0;
+    return 1;
+}
+
+int FP2_mul_frb(const BP_GROUP *group, FP2 *r, const FP2 *a, int i,
+                BN_CTX *ctx)
+{
+    FP2 *frb = NULL;
+    BN_CTX *new_ctx = NULL;
+    int ret = 0;
+
+    if (ctx == NULL)
+        if ((ctx = new_ctx = BN_CTX_new()) == NULL)
+            return 0;
+
+    if ((frb = FP2_new()) == NULL)
+        goto err;
+
+    /*
+     * Multiply by powers of residue powered to (p-1)/6.
+     */
+    if (i == 1) {
+        if (!FP2_mul(group, r, a, group->frb, ctx))
+            goto err;
+    }
+
+    if (i == 2) {
+        if (!FP2_sqr(group, frb, group->frb, ctx))
+            goto err;
+        if (!FP2_mul(group, r, a, frb, ctx))
+            goto err;
+    }
+
+    if (i == 3) {
+        if (!FP2_sqr(group, frb, group->frb, ctx))
+            goto err;
+        if (!FP2_mul(group, r, a, frb, ctx))
+            goto err;
+        if (!FP2_mul(group, r, a, group->frb, ctx))
+            goto err;
+    }
+
+    if (i == 4) {
+        if (!FP2_sqr(group, frb, group->frb, ctx))
+            goto err;
+        if (!FP2_sqr(group, frb, frb, ctx))
+            goto err;
+        if (!FP2_mul(group, r, a, frb, ctx))
+            goto err;
+    }
+
+    if (i == 5) {
+        if (!FP2_sqr(group, frb, group->frb, ctx))
+            goto err;
+        if (!FP2_sqr(group, frb, frb, ctx))
+            goto err;
+        if (!FP2_mul(group, r, a, frb, ctx))
+            goto err;
+        if (!FP2_mul(group, r, a, group->frb, ctx))
+            goto err;
+    }
+
+    ret = 1;
+
+ err:
+    FP2_free(frb);
+    BN_CTX_free(new_ctx);
+    return ret;
+}
+
+int FP2_mul(const BP_GROUP *group, FP2 *r, const FP2 *a, const FP2 *b,
+            BN_CTX *ctx)
+{
+    BIGNUM *t0, *t1, *t2, *t3, *t4;
+    BN_CTX *new_ctx = NULL;
+    int ret = 0;
+
+    if (ctx == NULL)
+        if ((ctx = new_ctx = BN_CTX_new()) == NULL)
+            return 0;
+
+    BN_CTX_start(ctx);
+    if ((t0 = BN_CTX_get(ctx)) == NULL || (t1 = BN_CTX_get(ctx)) == NULL
```

Comment:
Once BN_CTX_get fails, it continues to return NULL, so you can only check the return of the last invocation. You do need to initialize the pointers, though.